### ISSUE TITLE
SQLite plugin mapnik::loadMap() performance

### ISSUE THREAD

**Issue by jsimomaa** [mapnik/mapnik-support#84]
I've been hacking around with [jlnr/mapnik-jni](https://github.com/jlnr/mapnik-jni) bindings on Windows 10 out of curiosity and I have some questions about the performance of `mapnik::loadMap()` together with sqlite-plugin.

My workflow so far:
1. downloaded `estonia-latest.osm.pbf` (63 878 KB)
2. using `ogr2ogr` I converted the data to spatialite db `estonia.sqlite` (553 408 KB)
3. made necessary changes to [hotosm/tilemill-projects-windows/osm-sqlite](https://github.com/hotosm/tilemill-projects-windows/tree/master/osm-sqlite) such as renaming `extract.sqlite` to `estonia.sqlite` and added extents for the layers (this sped up the `mapnik::loadMap()` call quite a bit)

When looking at the database with `spatialite_gui` everything looks okay to me:

![](http://i.imgur.com/eq1TOUA.png)

Still the loading of the map takes up to few minutes on my system. I compiled Mapnik by using the help of [mapbox/windows-builds](https://github.com/mapbox/windows-builds) with `#MAPNIK_STATS`, `#MAPNIK_LOG` and `PACKAGEDEBUGSYMBOLS=1` enabled to get some statistics and debugging available.

Below are part of the prints of the `mapnik::loadMap()` call:

```
0.07ms (cpu 0.00ms)         | sqlite_resultset::execute_query SELECT * FROM (( SELECT OGC_FID, GEOMETRY, way_area AS area, COALESCE(landuse, leisure, "natural", amenity, tourism) AS type
  FROM planet_osm_polygon
  ORDER BY way_area DESC
) AS data) LIMIT 1
0.04ms (cpu 0.00ms)         | sqlite_resultset::execute_query PRAGMA table_info(planet_osm_polygon)
0.41ms (cpu 0.00ms)         | sqlite_resultset::execute_query SELECT pkid,xmin,xmax,ymin,ymax FROM "idx_planet_osm_polygon_GEOMETRY" LIMIT 1
0.63ms (cpu 1.00ms)         | sqlite_datasource::init(use_spatial_index)
7137.18ms (cpu 7138.00ms)   | sqlite_datasource::init
```

(see complete log [here](https://gist.github.com/jsimomaa/0c4630fca664969145500b4d15618b47))

Here we can see that the initialization of `sqlite_datasource::init` takes up to 7 seconds and as OSMBright style has 20 different layers the time stacks up. I debugged the function call with Visual Studio and found out the following:

![image](https://cloud.githubusercontent.com/assets/4503494/19564494/6d7691ba-96ec-11e6-90b9-b261a06664e8.png)

([See here for raw text](https://gist.github.com/jsimomaa/1f19f8d95297330a87a16405bb7e9c5e))

Here is the java code I'm using to load the map:

```
MapDefinition m = new MapDefinition();
m.loadMap("osm-sqlite.xml", false);
```

I'm just wondering if this is normal for using Mapnik with `sqlite.plugin` or not?

Thanks in advance!

**Comment by wilhelmberg** [comment]
@jsimomaa would you mind running another profiling session?
This time including CPU usage?
Like here: https://msdn.microsoft.com/en-us/library/dn971856.aspx

Would be interesting to see which method needs the most juice.

**Comment by jsimomaa** [comment]
@BergWerkGIS I tried to do the CPU usage profiling but it wasn't possible because the  Visual Studio solution builds a DLL and not an executable. Here are the options the CPU usage profiler gives me: 

![image](https://cloud.githubusercontent.com/assets/4503494/19570602/5e62f940-9703-11e6-99de-2a6306be7df1.png)

I may also need to clarify that I load `mapnik.dll` (and dependencies) via Java by calling

```
System.loadLibrary("mapnik.dll");
```

And the DLL is built in `Release` mode due to restriction in [here](https://github.com/mapbox/windows-builds/blob/master/settings.bat#L81).

The only way I can profile a release build is to use `Performance Explorer > Actions > Attach/Detach` and select the `javaw.exe` process which has loaded the DLL.

The system I'm using is

```
i7 6700HQ @ 2.60 GHz
32db RAM
512gb NVMe PM951 Samsung m.2 SSD
```

I'll attach the session I profiled in a zip-archive and a hopefully more informative screenshot below:

![image](https://cloud.githubusercontent.com/assets/4503494/19570515/07dca346-9703-11e6-89d2-c196c11f9bc1.png)

[PerformanceReport161020.zip](https://github.com/mapnik/mapnik-support/files/542398/PerformanceReport161020.zip)

Thank you for the feedback!

**Comment by jsimomaa** [comment]
> Still the loading of the map takes up to few minutes on my system.

I over-exaggerated this quite much as it seems from the screenshot above (~15 seconds). Although this is the case when I'm loading 8 maps concurrently (one for each thread of my CPU). The above profiling is for single `mapnik::loadMap()` function

**Comment by wilhelmberg** [comment]
> And the DLL is built in Release mode due to restriction in here.

Yeah, this is bad, but we had to do it, because creation of `pbd`s in `Debug` crashed VS when building.

> I tried to do the CPU usage profiling but it wasn't possible because the Visual Studio solution builds a DLL and not an executable. Here are the options the CPU usage profiler gives me:

How about selecting `Executable` and then passing in `javaw.exe`?
Could that work with your setup?

![image](https://cloud.githubusercontent.com/assets/4549888/19574868/60f966a4-970c-11e6-836b-e36a4d8d1921.png)

**Comment by jsimomaa** [comment]
Oh, I didn't even think about using the `javaw.exe` as the executable as the analysis target (I'm launching my java-code through IDE normally).

Okay, I tried with the following configuration

![image](https://cloud.githubusercontent.com/assets/4503494/19576998/09e891d4-971e-11e6-8cc8-eb9e1349435d.png)

But unfortunately the results of the CPU sampling seems the same as with the `Performance Explorer`

![image](https://cloud.githubusercontent.com/assets/4503494/19577043/320779dc-971e-11e6-8b42-3734adffad60.png)

![image](https://cloud.githubusercontent.com/assets/4503494/19577087/5df136c8-971e-11e6-8472-9bed7ca552c0.png)

I'll attach the exported report data here though

[javaw161020.zip](https://github.com/mapnik/mapnik-support/files/542837/javaw161020.zip)
